rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/RequestStrategy.kt

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.rsocket.kotlin
+
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.channels.*
+import kotlin.coroutines.*
+import kotlin.native.concurrent.*
+
+@SharedImmutable
+@ExperimentalStreamsApi
+private val DefaultStrategy: RequestStrategy = PrefetchStrategy(64, 16)
+
+@ExperimentalStreamsApi
+internal fun CoroutineContext.requestStrategy(): RequestStrategy.Element = (get(RequestStrategy) ?: DefaultStrategy).provide()
+
+@ExperimentalStreamsApi
+public interface RequestStrategy : CoroutineContext.Element {
+    override val key: CoroutineContext.Key<*> get() = Key
+
+    public fun provide(): Element
+
+    public interface Element {
+        public suspend fun firstRequest(): Int
+        public suspend fun nextRequest(): Int
+    }
+
+    public companion object Key : CoroutineContext.Key<RequestStrategy>
+}
+
+
+//request `requestSize` when `requestOn` elements left for collection
+//f.e. requestSize = 30, requestOn = 10, then first requestN will be 30, after 20 elements will be collected,
+//     new requestN for 30 elements will be sent so collect will be smooth
+@ExperimentalStreamsApi
+public class PrefetchStrategy(
+    private val requestSize: Int,
+    private val requestOn: Int,
+) : RequestStrategy {
+    init {
+        require(requestOn in 0 until requestSize) { "requestSize and requestOn should be in relation: requestSize > requestOn >= 0" }
+    }
+
+    override fun provide(): RequestStrategy.Element = Element(requestSize, requestOn)
+
+    private class Element(
+        private val requestSize: Int,
+        private val requestOn: Int,
+    ) : RequestStrategy.Element {
+        private val requested = atomic(requestSize)
```

Comment:
Just wondering why do we need atomic here?

Reply:
it needed to allow mutation on K/N if we freeze. need to check, may be it's really not needed, here.
On K/JVM and K/JS it's not needed, as functions inside `RequestStrategy.Element` will never be called concurrently

Reply:
yes, please double-check. In general, it looks like all the requests will be done within the same context, hence the context should do all the stuff to keep seriality, hence no atomic is needed

Reply:
atomic removed, not needed, thx for pointing!